File: cpp_engine/src/gpu.cpp
#include "gpu.h"
#include "config.h"
#include <array>
#include <iostream>
#include <sstream>
#include <unordered_map>
#include <unordered_set>
#include <set>
#include <algorithm>
#include <thread>
#include <chrono>

#ifdef _WIN32
#include <windows.h>
#include <dxgi.h>
#pragma comment(lib, "dxgi.lib")
#endif

namespace engine {

struct EncoderDef {
    std::string id;
    std::string vendor;
    std::string hwaccel;
    std::string label;
};

static const std::vector<EncoderDef> ALL_HARDWARE_ENCODERS = {
    {"h264_nvenc", "NVIDIA", "cuda", "NVIDIA NVENC H.264"},
    {"hevc_nvenc", "NVIDIA", "cuda", "NVIDIA NVENC HEVC"},
    {"av1_nvenc",  "NVIDIA", "cuda", "NVIDIA NVENC AV1"},
    {"h264_amf",   "AMD",    "amf",  "AMD AMF H.264"},
    {"hevc_amf",   "AMD",    "amf",  "AMD AMF HEVC"},
    {"av1_amf",    "AMD",    "amf",  "AMD AMF AV1"},
    {"h264_qsv",   "Intel",  "qsv",  "Intel Quick Sync H.264"},
    {"hevc_qsv",   "Intel",  "qsv",  "Intel Quick Sync HEVC"},
    {"av1_qsv",    "Intel",  "qsv",  "Intel Quick Sync AV1"},
};

static const std::vector<std::array<std::string,3>> RECOMMENDED_PRIORITY = {
    {"h264_nvenc", "NVIDIA", "cuda"},
    {"h264_amf",   "AMD",    "amf"},
    {"h264_qsv",   "Intel",  "qsv"},
};

static std::vector<std::string> get_system_gpu_names() {
    std::vector<std::string> gpus;
#ifdef _WIN32
    IDXGIFactory* factory = nullptr;
    if (SUCCEEDED(CreateDXGIFactory(__uuidof(IDXGIFactory), (void**)&factory))) {
        IDXGIAdapter* adapter = nullptr;
        for (UINT i = 0; factory->EnumAdapters(i, &adapter) != DXGI_ERROR_NOT_FOUND; ++i) {
            DXGI_ADAPTER_DESC desc;
            if (SUCCEEDED(adapter->GetDesc(&desc))) {
                wchar_t wname[128] = {};
                char cname[128] = {};
                size_t count = 0;
                wcstombs_s(&count, cname, sizeof(cname), desc.Description, _TRUNCATE);
                gpus.push_back(cname);
            }
            adapter->Release();
        }
        factory->Release();
    }
#elif defined(__linux__)
    std::array<char, 4096> buf;
    FILE* pipe = popen("lspci", "r");
    if (pipe) {
        while (fgets(buf.data(), buf.size(), pipe)) {
            std::string line(buf.data());
            auto lower = line;
            std::transform(lower.begin(), lower.end(), lower.begin(), ::tolower);
            if (lower.find("vga") != std::string::npos ||
                lower.find("3d") != std::string::npos ||
                lower.find("display") != std::string::npos) {
                auto pos = line.find(": ", 0);
                if (pos != std::string::npos) {
                    auto name = line.substr(pos + 2);
                    while (!name.empty() && (name.back() == '\n' || name.back() == '\r'))
                        name.pop_back();
                    gpus.push_back(name);
                }
            }
        }
        pclose(pipe);
    }
#elif defined(__APPLE__)
    std::array<char, 4096> buf;
    FILE* pipe = popen("system_profiler SPDisplaysDataType", "r");
    if (pipe) {
        while (fgets(buf.data(), buf.size(), pipe)) {
            std::string line(buf.data());
            if (line.find("Chipset Model") != std::string::npos ||
                line.find("Chip Model") != std::string::npos) {
                auto pos = line.find(": ", 0);
                if (pos != std::string::npos) {
                    auto name = line.substr(pos + 2);
                    while (!name.empty() && (name.back() == '\n' || name.back() == '\r'))
                        name.pop_back();
                    gpus.push_back(name);
                }
            }
        }
        pclose(pipe);
    }
#endif
    return gpus;
}

static std::unordered_set<std::string> probe_ffmpeg_encoders() {
    std::unordered_set<std::string> encoders;
    auto ffmpeg = find_binary("ffmpeg");
    std::string cmd = ffmpeg + " -hide_banner -encoders 2>&1";

    FILE* pipe = _popen(cmd.c_str(), "r");
    if (!pipe) return encoders;

    char buf[4096];
    while (fgets(buf, sizeof(buf), pipe)) {
        std::string line(buf);
        std::istringstream iss(line);
        std::vector<std::string> parts;
        std::string part;
        while (iss >> part) parts.push_back(part);
        if (parts.size() >= 2) {
            encoders.insert(parts[1]);
        }
    }
    _pclose(pipe);
    return encoders;
}

static bool encoder_works_cached(const std::string& encoder) {
    static std::unordered_map<std::string, bool> cache;
    auto it = cache.find(encoder);
    if (it != cache.end()) return it->second;

    auto ffmpeg = find_binary("ffmpeg");
    std::string cmd = ffmpeg + " -v error -nostdin -f lavfi -i color=c=black:s=256x256:r=1:d=0.1"
                      " -frames:v 1 -c:v " + encoder + " -pix_fmt yuv420p -f null - 2>&1";

    FILE* pipe = _popen(cmd.c_str(), "r");
    bool works = false;
    if (pipe) {
        char buf[256];
        while (fgets(buf, sizeof(buf), pipe)) {}
        int rc = _pclose(pipe);
        works = (rc == 0);
    }

    std::cerr << "[gpu] Testing encoder: " << encoder << " -> " << (works ? "WORKS" : "FAILED") << std::endl;
    cache[encoder] = works;
    return works;
}

static std::string to_lower_str(const std::string& s) {
    std::string r = s;
    std::transform(r.begin(), r.end(), r.begin(), ::tolower);
    return r;
}

GpuDetectionResult detect_gpu() {
    std::cerr << "[gpu] Running GPU detection..." << std::endl;

    auto compiled = probe_ffmpeg_encoders();
    auto gpus = get_system_gpu_names();

    std::vector<EncoderInfo> available;
    for (auto& enc : ALL_HARDWARE_ENCODERS) {
        if (compiled.count(enc.id) && encoder_works_cached(enc.id)) {
            std::string gpu_name;
            for (auto& g : gpus) {
                if (to_lower_str(g).find(to_lower_str(enc.vendor)) != std::string::npos) {
                    gpu_name = g;
                    break;
                }
            }
            std::string label = enc.label;
            if (!gpu_name.empty()) label += " (" + gpu_name + ")";
            available.push_back({enc.id, enc.vendor, enc.hwaccel, label});
        }
    }

    std::optional<EncoderInfo> best_encoder;
    for (auto& rp : RECOMMENDED_PRIORITY) {
        for (auto& e : available) {
            if (e.id == rp[0]) {
                best_encoder = e;
                break;
            }
        }
        if (best_encoder) break;
    }

    nlohmann::json all_encoders_json = nlohmann::json::array();
    if (best_encoder) {
        all_encoders_json.push_back({
            {"id", best_encoder->id},
            {"vendor", best_encoder->vendor},
            {"label", best_encoder->label + " (Recommended)"}
        });
    }
    for (auto& e : available) {
        if (!best_encoder || e.id != best_encoder->id) {
            all_encoders_json.push_back({
                {"id", e.id},
                {"vendor", e.vendor},
                {"label", e.label}
            });
        }
    }
    all_encoders_json.push_back({
        {"id", "libx264"},
        {"vendor", "CPU"},
        {"label", "Software H.264 (CPU fallback)"}
    });

    std::string gpu_name = gpus.empty() ? "" : gpus[0];

    std::cerr << "[gpu] Best encoder: " << (best_encoder ? best_encoder->id : "none") << std::endl;

    GpuDetectionResult result;
    if (best_encoder) {
        std::string display_name = gpu_name;
        for (auto& g : gpus) {
            if (to_lower_str(g).find(to_lower_str(best_encoder->vendor)) != std::string::npos) {
                display_name = g;
                break;
            }
        }
        if (display_name.empty()) display_name = best_encoder->vendor + " GPU";

        result.available = true;
        result.encoder = best_encoder->id;
        result.vendor = best_encoder->vendor;
        result.hwaccel = best_encoder->hwaccel;
        result.name = display_name;
        result.message = "Detected: " + display_name + " (" + best_encoder->id + ")";
    } else {
        result.available = false;
        result.message = "No hardware encoder found" + (gpu_name.empty() ? "" : " (GPU: " + gpu_name + ")");
        result.name = gpu_name.empty() ? std::nullopt : std::make_optional(gpu_name);
    }

    for (auto& e : all_encoders_json) {
        EncoderInfo info;
        info.id = e["id"].get<std::string>();
        info.vendor = e["vendor"].get<std::string>();
        info.label = e["label"].get<std::string>();
        result.all_encoders.push_back(info);
    }

    return result;
}

std::vector<std::string> get_hwaccel_for_encoder(const std::string& encoder) {
    for (auto& enc : ALL_HARDWARE_ENCODERS) {
        if (enc.id == encoder) {
            return {"-hwaccel", enc.hwaccel};
        }
    }
    if (encoder.size() >= 6 && encoder.substr(encoder.size() - 6) == "_nvenc") return {"-hwaccel", "cuda"};
    if (encoder.size() >= 4 && encoder.substr(encoder.size() - 4) == "_amf")  return {"-hwaccel", "amf"};
    if (encoder.size() >= 4 && encoder.substr(encoder.size() - 4) == "_qsv")  return {"-hwaccel", "qsv"};
    return {};
}

std::vector<std::string> get_hwaccel_args(bool use_gpu, const std::optional<std::string>& encoder) {
    if (!use_gpu) return {};
    auto info = detect_gpu();
    if (!info.available || !info.hwaccel) return {};
    if (encoder && encoder->size() >= 6 && encoder->substr(encoder->size() - 6) == "_nvenc") {
        return {"-hwaccel", "cuda", "-hwaccel_output_format", "cuda"};
    }
    return {"-hwaccel", *info.hwaccel};
}

std::vector<std::string> get_cuda_decode_args(const std::string& encoder) {
    if (encoder.size() < 6 || encoder.substr(encoder.size() - 6) != "_nvenc") return {};
    return {"-hwaccel", "cuda", "-hwaccel_output_format", "cuda"};
}

std::string get_cuda_scale_filter(int width, int max_width) {
    if (max_width <= 0 || width <= max_width) return "";
    int target_w = std::max(2, (max_width / 2) * 2);
    return "scale_cuda=" + std::to_string(target_w) + ":-2";
}

std::string get_video_encoder(bool use_gpu, const std::string& fallback,
                              const std::string& preferred_encoder,
                              const std::optional<std::string>& output_format) {
    if (use_gpu) {
        if (fallback != "libx264") {
            if (fallback == "gif") return fallback;
            throw std::runtime_error("This output codec requires CPU encoding. Choose MP4/MKV for GPU encoding, or select CPU in Settings.");
        }
        for (auto& rp : RECOMMENDED_PRIORITY) {
            if (encoder_works_cached(rp[0])) return rp[0];
        }
        throw std::runtime_error("No working GPU video encoder is available. Select CPU in Settings to use software encoding.");
    }
    if (preferred_encoder.empty() || preferred_encoder == "libx264") return fallback;

    std::unordered_set<std::string> known;
    for (auto& e : ALL_HARDWARE_ENCODERS) known.insert(e.id);
    if (!known.count(preferred_encoder)) throw std::runtime_error("Unknown preferred encoder");

    bool compatible = false;
    if (preferred_encoder.substr(0, 5) == "h264_" && fallback == "libx264") compatible = true;
    if (preferred_encoder.substr(0, 5) == "hevc_" && output_format &&
        (std::set<std::string>{"mp4","mkv","mov","m4v"}.count(*output_format))) compatible = true;
    if (preferred_encoder.substr(0, 4) == "av1_" && output_format &&
        (std::set<std::string>{"mp4","mkv","webm"}.count(*output_format))) compatible = true;
    if (!compatible) {
        throw std::runtime_error("The selected GPU encoder is incompatible with this output format.");
    }
    if (encoder_works_cached(preferred_encoder)) return preferred_encoder;
    throw std::runtime_error("The selected GPU encoder (" + preferred_encoder + ") is unavailable.");
}

std::vector<std::string> video_encoding_args(const std::string& encoder, std::optional<int> quality) {
    std::vector<std::string> args = {"-c:v", encoder};
    if (encoder.size() >= 6 && encoder.substr(encoder.size() - 6) == "_nvenc") {
        args.insert(args.end(), {"-preset", "p4"});
        if (quality) {
            args.insert(args.end(), {"-rc", "vbr", "-cq", std::to_string(*quality), "-b:v", "0"});
        }
    } else if (encoder.size() >= 4 && encoder.substr(encoder.size() - 4) == "_amf") {
        args.insert(args.end(), {"-quality", "balanced"});
        if (quality) {
            args.insert(args.end(), {"-rc", "cqp", "-qp_i", std::to_string(*quality), "-qp_p", std::to_string(*quality)});
        }
    } else if (encoder.size() >= 4 && encoder.substr(encoder.size() - 4) == "_qsv") {
        args.insert(args.end(), {"-preset", "medium"});
        if (quality) {
            args.insert(args.end(), {"-global_quality", std::to_string(*quality)});
        }
    } else if (encoder == "libx264" || encoder == "libx265") {
        args.insert(args.end(), {"-preset", "medium"});
        if (quality) {
            args.insert(args.end(), {"-crf", std::to_string(*quality)});
        }
    }
    return args;
}

}

File: PyEngine/core/gpu.py
"""GPU hardware encoder detection via ffmpeg."""
import subprocess
import sys
from functools import lru_cache
from typing import Optional

from PyEngine.core.config import find_binary

# All known hardware encoders with their vendor and hwaccel method.
_ALL_HARDWARE_ENCODERS = [
    ("h264_nvenc", "NVIDIA", "cuda", "NVIDIA NVENC H.264"),
    ("hevc_nvenc", "NVIDIA", "cuda", "NVIDIA NVENC HEVC"),
    ("av1_nvenc", "NVIDIA", "cuda", "NVIDIA NVENC AV1"),
    ("h264_amf", "AMD", "amf", "AMD AMF H.264"),
    ("hevc_amf", "AMD", "amf", "AMD AMF HEVC"),
    ("av1_amf", "AMD", "amf", "AMD AMF AV1"),
    ("h264_qsv", "Intel", "qsv", "Intel Quick Sync H.264"),
    ("hevc_qsv", "Intel", "qsv", "Intel Quick Sync HEVC"),
    ("av1_qsv", "Intel", "qsv", "Intel Quick Sync AV1"),
]

# Recommended priority: NVIDIA > AMD > Intel, H.264 first.
_RECOMMENDED_PRIORITY = [
    ("h264_nvenc", "NVIDIA", "cuda"),
    ("h264_amf", "AMD", "amf"),
    ("h264_qsv", "Intel", "qsv"),
]

_CPU_ENCODERS = [
    ("libx264", "CPU", None, "Software H.264 (libx264)"),
]


def _get_system_gpu_names() -> list:
    """Return a list of GPU names on this system."""
    gpus = []
    try:
        if sys.platform == "win32":
            result = subprocess.run(
                [
                    "powershell", "-NoProfile", "-Command",
                    "Get-CimInstance Win32_VideoController | Select-Object -ExpandProperty Name",
                ],
                capture_output=True, text=True, timeout=5,
                creationflags=subprocess.CREATE_NO_WINDOW,
            )
            for line in result.stdout.strip().splitlines():
                name = line.strip()
                if name:
                    gpus.append(name)
        elif sys.platform == "linux":
            result = subprocess.run(["lspci"], capture_output=True, text=True, timeout=5)
            for line in result.stdout.splitlines():
                lower = line.lower()
                if "vga" in lower or "3d" in lower or "display" in lower:
                    parts = line.split(": ", 1)
                    if len(parts) > 1:
                        gpus.append(parts[1].strip())
        elif sys.platform == "darwin":
            result = subprocess.run(
                ["system_profiler", "SPDisplaysDataType"],
                capture_output=True, text=True, timeout=5,
            )
            for line in result.stdout.splitlines():
                if "Chipset Model" in line or "Chip Model" in line:
                    parts = line.split(": ", 1)
                    if len(parts) > 1:
                        gpus.append(parts[1].strip())
    except Exception:
        pass
    return gpus


def _probe_ffmpeg_encoders() -> set:
    """Return the set of encoder names supported by the bundled ffmpeg."""
    ffmpeg = find_binary("ffmpeg")
    try:
        result = subprocess.run(
            [ffmpeg, "-hide_banner", "-encoders"],
            capture_output=True, text=True, timeout=10,
            creationflags=subprocess.CREATE_NO_WINDOW if sys.platform == "win32" else 0,
        )
        encoders = set()
        for line in result.stdout.splitlines():
            parts = line.split()
            if len(parts) >= 2:
                encoders.add(parts[1])
        return encoders
    except Exception:
        return set()


@lru_cache(maxsize=16)
def _encoder_works(encoder: str) -> bool:
    """Test if an encoder actually works by encoding a single test frame."""
    try:
        result = subprocess.run(
            [find_binary("ffmpeg"), "-v", "error", "-nostdin", "-f", "lavfi",
             "-i", "color=c=black:s=256x256:r=1:d=0.1",
             "-frames:v", "1",
             "-c:v", encoder, "-pix_fmt", "yuv420p",
             "-f", "null", "-"],
            capture_output=True, timeout=8,
            creationflags=subprocess.CREATE_NO_WINDOW if sys.platform == "win32" else 0,
        )
        print(f"[gpu] Testing encoder: {encoder} -> {'WORKS' if result.returncode == 0 else 'FAILED'}", file=sys.stderr, flush=True)
        return result.returncode == 0
    except (OSError, subprocess.TimeoutExpired):
        print(f"[gpu] Testing encoder: {encoder} -> FAILED (exception)", file=sys.stderr, flush=True)
        return False


@lru_cache(maxsize=None)
def detect_gpu() -> dict:
    """Detect the best GPU and all available encoders.

    Returns a dict with:
        available: bool - whether a hardware encoder was found
        encoder: str|null - best ffmpeg encoder name
        vendor: str|null - best GPU vendor
        hwaccel: str|null - best hwaccel method
        name: str|null - human-readable best GPU name
        message: str - status message
        all_encoders: list - all working encoders [{"id", "vendor", "label"}]
    """
    print(f"[gpu] Running GPU detection...", file=sys.stderr, flush=True)
    compiled = _probe_ffmpeg_encoders()
    gpus = _get_system_gpu_names()

    # Find all working hardware encoders
    available = []
    for enc_id, vendor, hwaccel, label in _ALL_HARDWARE_ENCODERS:
        if enc_id in compiled and _encoder_works(enc_id):
            # Match encoder to a GPU name if possible
            gpu_name = next((g for g in gpus if vendor.lower() in g.lower()), None)
            available.append({
                "id": enc_id,
                "vendor": vendor,
                "hwaccel": hwaccel,
                "label": f"{label}" + (f" ({gpu_name})" if gpu_name else ""),
            })

    # Find the recommended (best) encoder
    best_encoder = None
    for enc_id, vendor, hwaccel in _RECOMMENDED_PRIORITY:
        if any(e["id"] == enc_id for e in available):
            best_encoder = (enc_id, vendor, hwaccel)
            break

    # Build the full encoder list: recommended first, then others, then CPU
    all_encoders = []
    if best_encoder:
        be_id, be_vendor, be_hwaccel = best_encoder
        be_info = next((e for e in available if e["id"] == be_id), None)
        be_label = be_info["label"] if be_info else be_id
        all_encoders.append({"id": be_id, "vendor": be_vendor, "label": f"{be_label} (Recommended)"})

    for e in available:
        if e["id"] != (best_encoder[0] if best_encoder else None):
            all_encoders.append({"id": e["id"], "vendor": e["vendor"], "label": e["label"]})

    all_encoders.append({"id": "libx264", "vendor": "CPU", "label": "Software H.264 (CPU fallback)"})

    # Pick best GPU name
    gpu_name = gpus[0] if gpus else None

    print(f"[gpu] Best encoder: {best_encoder}", file=sys.stderr, flush=True)
    print(f"[gpu] All working encoders: {[e['id'] for e in all_encoders]}", file=sys.stderr, flush=True)

    if best_encoder:
        be_id, be_vendor, be_hwaccel = best_encoder
        display_name = next((g for g in gpus if be_vendor.lower() in g.lower()), None) or f"{be_vendor} GPU"
        return {
            "available": True,
            "encoder": be_id,
            "vendor": be_vendor,
            "hwaccel": be_hwaccel,
            "name": display_name,
            "message": f"Detected: {display_name} ({be_id})",
            "all_encoders": all_encoders,
        }

    return {
        "available": False,
        "encoder": None,
        "vendor": None,
        "hwaccel": "",
        "name": gpu_name,
        "message": "No hardware encoder found" + (f" (GPU: {gpu_name})" if gpu_name else ""),
        "all_encoders": all_encoders,
    }


def get_hwaccel_for_encoder(encoder: str) -> list:
    """Return ffmpeg hwaccel args for a specific hardware encoder."""
    for enc_id, vendor, hwaccel, label in _ALL_HARDWARE_ENCODERS:
        if enc_id == encoder:
            return ["-hwaccel", hwaccel]
    # Match by suffix
    if encoder.endswith("_nvenc"):
        return ["-hwaccel", "cuda"]
    elif encoder.endswith("_amf"):
        return ["-hwaccel", "amf"]
    elif encoder.endswith("_qsv"):
        return ["-hwaccel", "qsv"]
    return []


def get_hwaccel_args(use_gpu: bool, encoder: Optional[str] = None) -> list:
    """Return ffmpeg args for hardware acceleration when use_gpu is True.

    When encoder is an NVENC variant and use_gpu is True, uses
    -hwaccel cuda -hwaccel_output_format cuda so decoded frames stay on GPU.
    Otherwise falls back to -hwaccel <method> or empty list.
    """
    if not use_gpu:
        return []
    info = detect_gpu()
    if not info["available"] or not info["hwaccel"]:
        return []
    # Full CUDA pipeline: decode on GPU, keep frames in GPU memory
    if encoder and encoder.endswith("_nvenc"):
        return ["-hwaccel", "cuda", "-hwaccel_output_format", "cuda"]
    # Generic hwaccel (software decode, GPU encode)
    return ["-hwaccel", info["hwaccel"]]


def get_cuda_decode_args(encoder: str) -> list:
    """Return CUDA decode args (keeps frames on GPU) when using an NVENC encoder.

    Uses -hwaccel cuda -hwaccel_output_format cuda so decoded frames stay in GPU
    memory. Falls back gracefully if NVDEC is unavailable for the input codec.
    """
    if not encoder or not encoder.endswith("_nvenc"):
        return []
    return ["-hwaccel", "cuda", "-hwaccel_output_format", "cuda"]


def get_cuda_scale_filter(width: int, max_width: int) -> str:
    """Return a scale_cuda filter string if width needs downscaling, else empty string.

    Uses the GPU-resident format from -hwaccel_output_format cuda.
    """
    if max_width <= 0 or width <= max_width:
        return ""
    target_w = max(2, max_width // 2 * 2)
    return f"scale_cuda={target_w}:-2"


def get_video_encoder(use_gpu: bool, fallback: str = "libx264", preferred_encoder: Optional[str] = None,
                      output_format: Optional[str] = None) -> str:
    """Auto mode selects H.264; manual mode respects format compatibility."""
    if use_gpu:
        if fallback != "libx264":
            if fallback == "gif":
                return fallback
            raise ValueError("This output codec requires CPU encoding. Choose MP4/MKV for GPU encoding, or select CPU in Settings.")
        # Jobs need one working encoder, not a full inventory and GPU-name
        # subprocess. Stop at the first usable H.264 encoder.
        encoder = next((enc for enc, _, _ in _RECOMMENDED_PRIORITY if _encoder_works(enc)), None)
        if encoder:
            return encoder
        raise ValueError("No working GPU video encoder is available. Select CPU in Settings to use software encoding.")
    if not preferred_encoder or preferred_encoder == "libx264":
        return fallback
    known = {entry[0] for entry in _ALL_HARDWARE_ENCODERS}
    if preferred_encoder not in known:
        raise ValueError("Unknown preferred encoder")
    compatible = (
        (preferred_encoder.startswith("h264_") and fallback == "libx264")
        or (preferred_encoder.startswith("hevc_") and output_format in {"mp4", "mkv", "mov", "m4v"})
        or (preferred_encoder.startswith("av1_") and output_format in {"mp4", "mkv", "webm"})
    )
    if not compatible:
        raise ValueError("The selected GPU encoder is incompatible with this output format. Choose a compatible format or select CPU in Settings.")
    if _encoder_works(preferred_encoder):
        return preferred_encoder
    raise ValueError(f"The selected GPU encoder ({preferred_encoder}) is unavailable. Choose another encoder in Settings.")


def video_encoding_args(encoder: str, quality: Optional[int] = None) -> list:
    """Select preset and quality options supported by the chosen encoder."""
    args = ["-c:v", encoder]
    if encoder.endswith("_nvenc"):
        args += ["-preset", "p4"]
        if quality is not None:
            args += ["-rc", "vbr", "-cq", str(quality), "-b:v", "0"]
    elif encoder.endswith("_amf"):
        args += ["-quality", "balanced"]
        if quality is not None:
            args += ["-rc", "cqp", "-qp_i", str(quality), "-qp_p", str(quality)]
    elif encoder.endswith("_qsv"):
        args += ["-preset", "medium"]
        if quality is not None:
            args += ["-global_quality", str(quality)]
    elif encoder in ("libx264", "libx265"):
        args += ["-preset", "medium"]
        if quality is not None:
            args += ["-crf", str(quality)]
    return args


def resolve_encoder_selection(use_gpu, preferred_encoder="", selected_gpu=""):
    """Normalize JSON settings to the workers' auto/explicit encoder contract."""
    known = {"", "libx264", *(entry[0] for entry in _ALL_HARDWARE_ENCODERS)}
    if not isinstance(use_gpu, bool):
        raise ValueError("use_gpu must be a boolean")
    for encoder in (preferred_encoder, selected_gpu):
        if not isinstance(encoder, str) or encoder not in known:
            raise ValueError("Invalid GPU encoder")
    if use_gpu and selected_gpu:
        return False, selected_gpu
    return use_gpu, preferred_encoder
